### PR TITLE
Hold my context!

### PR DESCRIPTION
**Pull request by bajtos**

### CODE REVIEW

File: 2020-hold-my-context.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# Hold my context!
+
+☝️  _the presentation title_☝️
+
+## Short summary for display
+
+Preserving contextual data across asynchronous calls have been always tricky
+in Node.js land. Traditional runtimes like C and Java can utilize thread-local
+storage, which obviously does not work for Node's single-threaded event loop.
```
Comment (BethGriggs) [inline 2020-hold-my-context.md]:
```suggestion
storage, which does not work for Node's single-threaded event loop.
```

File: 2020-hold-my-context.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+Preserving contextual data across asynchronous calls have been always tricky
+in Node.js land. For example, how to pass information about the current user
+or a correlation (transaction) id from the code processing incoming HTTP
+requests to the code calling a backend web service? Traditional runtimes like
+C and Java can utilize thread-local storage, which does not work for Node's
+single-threaded event loop.
+
+The talk will describe different solutions people have been using (and
+abandoning) over the past 6+ years, dive into history of domains and async
+hooks, explain what's so difficult about context keeping, and finally offer
```
Comment (sam-github) [inline 2020-hold-my-context.md]:
```suggestion
hooks, explain what's so difficult about context keeping, and finally offer the
```

File: 2020-hold-my-context.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+bright future that's coming with the recently introduced `AsyncLocalStorage` API.
```
Comment (sam-github) [inline 2020-hold-my-context.md:17]: I don't know if Vladimir thinks the future is bright! https://github.com/nodejs/TSC/issues/807

File: 2020-hold-my-context.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+## Talk description and abstract
+
+_Structure, table of contents and main content points._
+
+1. Why would anyone need to preserve a context? Start with an example explaining
+  the problem faced by server developers: how to forward correlaction-id from
```
Comment (sam-github) [inline 2020-hold-my-context.md]:
```suggestion
  the problem faced by server developers: how to forward correlation-id from
```

File: 2020-hold-my-context.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+2. Solution 1: core `domains`. While they can track async transitions and
+  continuations, they are not meant to store custom data.
+
+3. Solution 2: `continuation-local-storage`. The go-to solution during Node.js
+   0.10 days. Unfortunately, it often drops the context and requires
+   applications to cooperate, which often means fixing dependencies 😱.
+
+4. Explain why is it difficult to preserve the context, a.k.a. "connection pool
+   problem". Talk about our own experience with issues caused by MongoDB driver.
+   Mention other use cases affected by this problem (APM tools like NewRelic).
```
Comment (sam-github) [inline 2020-hold-my-context.md]: Promises are a huge problem as well. Google's native promises have support for async hooks, forced to be added by Node.js, but thirdparty promises don't, so any non-native thenable breaks the async continuation. Apparently,  its particularly common for DB drivers.

Comment (sam-github) [inline 2020-hold-my-context.md]: Promises are a huge problem as well. Google's native promises have support for async hooks, forced to be added by Node.js, but thirdparty promises don't, so any non-native thenable breaks the async continuation. Apparently,  its particularly common for DB drivers.

**Review by sam-github — COMMENTED** [review]
Promises are a huge problem as well. Google's native promises have support for async hooks, forced to be added by Node.js, but thirdparty promises don't, so any non-native thenable breaks the async continuation. Apparently,  its particularly common for DB drivers.

looks like an interesting talk to me

What about applying for the Austin NodeConf with it as well?

File: 2020-hold-my-context.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
+# Hold my context!
+
+☝️  _the presentation title_☝️
+
+## Short summary for display
+
+Preserving contextual data across asynchronous calls have been always tricky
```
Comment (joesepi) [inline 2020-hold-my-context.md]: have should be has. "preserving contextual data ... has been"

**Review by joesepi — APPROVED** [review]
LGTM with one small nit. Good luck!

File: 2020-hold-my-context.md (hunk 3 above)
Comment (bajtos) [reply 2020-hold-my-context.md:17]: Good point. My impression was that https://github.com/nodejs/node/pull/26540 is on a good way to get landed soon. I'll mention in the abstract that the final message of my talk my change depending on the outcome of current discussions in Node.js core.

File: 2020-hold-my-context.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  world (C/Java/.NET).
+
+2. Solution 1: core `domains`. While they can track async transitions and
+  continuations, they are not meant to store custom data.
+
+3. Solution 2: `continuation-local-storage`. The go-to solution during Node.js
+   0.10 days. Unfortunately, it often drops the context and requires
+   applications to cooperate, which often means fixing dependencies 😱.
+
+4. Explain why is it difficult to preserve the context (user-land promises,
```
Comment (dhmlau) [inline 2020-hold-my-context.md]: nitpick: 
`Explain why is it difficult` -> `Explain why it is difficult....`

File: 2020-hold-my-context.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+   (APM tools like NewRelic).
+
+5. Solution 3: explicit context passing in function arguments. Very reliable
+   if you remember to always pass the context in function calls, but also
+   rather tedious.
+
+6. Introduce the idea of `async_hooks`, talk a bit about the history of the
+   implementation (`AsyncListener`, `async_wrap`).
+
+7. Solution 3: `cls-hooked` (Node.js 4.7+).  Works reasonably well, but
```
Comment (dhmlau) [inline 2020-hold-my-context.md]: Should be `Solution 4`?

File: 2020-hold-my-context.md [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+9. Using the example presented at the beginning, show how to use `AsyncLocalStorage`
+   API to implement forwarding of correlation-id values from incoming request
+   headers to the outgoing request headers when calling a web service.
+
+10. Conclusion. Availability of `AsyncLocalStorage` API, known limitations, etc.
+
+Please note that `AsyncLocalStorage` has not yet landed at the time of writing
+this proposal. The last three points of my talk may may slightly change to
+accomodate for the actual API the Node.js core team has settled on.
```
Comment (dhmlau) [inline 2020-hold-my-context.md]: Typo: `accomodate` -> `accommodate`

**Review by dhmlau — APPROVED** [review]
I don't have much technical knowledge on this, but the proposal and the flow LGTM. 
A few suggested changes on typos and grammar.

**Comment by bajtos** [comment]
Thank you all for the suggestions, I really appreciate your feedback 🙇 ❤️

**Pull request merged**